File: clients/typescript/test/unit/frames.test.ts
import { describe, expect, test } from "vitest";

import { parseServerFrame, subscribeFrame } from "../../src/frames.js";

describe("parseServerFrame", () => {
  test("parses a known event frame (conformance item 4/5 fields)", () => {
    const text = JSON.stringify({
      type: "event",
      id: "00000000-0000-0000-0000-000000000000",
      seq: 7,
      ts_ms: 1_751_760_000_000,
      channel: "github.push",
      payload_b64: "AA==",
    });
    const frame = parseServerFrame(text);
    expect(frame).toEqual({
      type: "event",
      id: "00000000-0000-0000-0000-000000000000",
      seq: 7,
      ts_ms: 1_751_760_000_000,
      channel: "github.push",
      payload_b64: "AA==",
    });
  });

  test("skips unknown frame type (conformance item 10)", () => {
    expect(parseServerFrame(JSON.stringify({ type: "quantum_flux", foo: 1 }))).toBeNull();
    expect(parseServerFrame("not json at all")).toBeNull();
    expect(parseServerFrame(JSON.stringify([1, 2, 3]))).toBeNull();
  });

  test("tolerates unknown extra fields on a known frame (conformance item 10)", () => {
    const text = JSON.stringify({ type: "welcome", name: "p", future_field: { nested: true } });
    expect(parseServerFrame(text)).toEqual({ type: "welcome", name: "p" });
  });

  test("rejects a known frame missing a required field", () => {
    // seq missing => not a usable event.
    expect(
      parseServerFrame(
        JSON.stringify({ type: "event", id: "x", ts_ms: 1, channel: "c", payload_b64: "AA==" }),
      ),
    ).toBeNull();
  });

  test("parses replayed / replay_gap / lagged control frames", () => {
    expect(parseServerFrame(JSON.stringify({ type: "replayed", through_seq: 128 }))).toEqual({
      type: "replayed",
      through_seq: 128,
    });
    expect(
      parseServerFrame(JSON.stringify({ type: "replay_gap", from_seq: 10, earliest_seq: 57 })),
    ).toEqual({ type: "replay_gap", from_seq: 10, earliest_seq: 57 });
    expect(parseServerFrame(JSON.stringify({ type: "lagged", dropped: 42 }))).toEqual({
      type: "lagged",
      dropped: 42,
    });
  });

  test("parses closing with both reasons; rejects an unknown reason", () => {
    expect(parseServerFrame(JSON.stringify({ type: "closing", reason: "shutdown" }))).toEqual({
      type: "closing",
      reason: "shutdown",
    });
    expect(parseServerFrame(JSON.stringify({ type: "closing", reason: "revoked" }))).toEqual({
      type: "closing",
      reason: "revoked",
    });
    expect(parseServerFrame(JSON.stringify({ type: "closing", reason: "aliens" }))).toBeNull();
  });

  test("parses ok / error / pong", () => {
    expect(parseServerFrame(JSON.stringify({ type: "ok", op: "subscribe" }))).toEqual({
      type: "ok",
      op: "subscribe",
    });
    expect(
      parseServerFrame(JSON.stringify({ type: "error", op: "replay", msg: "not enabled" })),
    ).toEqual({ type: "error", op: "replay", msg: "not enabled" });
    expect(parseServerFrame(JSON.stringify({ type: "pong" }))).toEqual({ type: "pong" });
  });
});

describe("subscribeFrame", () => {
  test("attaches replay.after_seq = cursor when given (conformance item 3)", () => {
    expect(subscribeFrame(["github.>"], 128)).toEqual({
      type: "subscribe",
      patterns: ["github.>"],
      replay: { after_seq: 128 },
    });
  });

  test("omits replay for live-only", () => {
    const frame = subscribeFrame(["github.>"], undefined);
    expect(frame).toEqual({ type: "subscribe", patterns: ["github.>"] });
    expect("replay" in frame).toBe(false);
  });

  test("resume cursor 0 still sends replay (cold-start replay from retention)", () => {
    expect(subscribeFrame(["a.>"], 0)).toEqual({
      type: "subscribe",
      patterns: ["a.>"],
      replay: { after_seq: 0 },
    });
  });
});

File: clients/go/frames_test.go
package whdrsub

import (
	"testing"
)

func TestParseKnownEventFrame(t *testing.T) {
	text := []byte(`{"type":"event","id":"00000000-0000-0000-0000-000000000000",
		"seq":7,"ts_ms":1751760000000,"channel":"github.push","payload_b64":"aGVsbG8="}`)
	frame, err := parseFrame(text)
	if err != nil {
		t.Fatalf("parseFrame: %v", err)
	}
	ev, ok := frame.(EventFrame)
	if !ok {
		t.Fatalf("expected EventFrame, got %T", frame)
	}
	if ev.Seq != 7 || ev.Channel != "github.push" {
		t.Fatalf("bad event: %+v", ev)
	}
	body, err := ev.Payload()
	if err != nil {
		t.Fatalf("Payload: %v", err)
	}
	if string(body) != "hello" {
		t.Fatalf("payload = %q, want hello", body)
	}
}

// Conformance item 10: unknown "type" values are ignored (nil, nil).
func TestSkipsUnknownFrameType(t *testing.T) {
	for _, text := range []string{
		`{"type":"quantum_flux","foo":1}`,
		`not json at all`,
		`{"no_type":true}`,
	} {
		frame, err := parseFrame([]byte(text))
		if err != nil {
			t.Fatalf("parseFrame(%q) err = %v", text, err)
		}
		if frame != nil {
			t.Fatalf("parseFrame(%q) = %v, want nil", text, frame)
		}
	}
}

// Conformance item 10: unknown object fields on a known frame are ignored.
func TestToleratesUnknownFieldsOnKnownFrame(t *testing.T) {
	text := []byte(`{"type":"welcome","name":"p","future_field":{"nested":true}}`)
	frame, err := parseFrame(text)
	if err != nil {
		t.Fatalf("parseFrame: %v", err)
	}
	w, ok := frame.(WelcomeFrame)
	if !ok {
		t.Fatalf("expected WelcomeFrame, got %T", frame)
	}
	if w.Name != "p" {
		t.Fatalf("name = %q, want p", w.Name)
	}
}

func TestParseAllServerFrameTypes(t *testing.T) {
	cases := []struct {
		text string
		want Frame
	}{
		{`{"type":"welcome","name":"sub-a"}`, WelcomeFrame{Name: "sub-a"}},
		{`{"type":"ok","op":"subscribe"}`, OkFrame{Op: "subscribe"}},
		{`{"type":"error","op":"replay","msg":"durability disabled"}`, ErrorFrame{Op: "replay", Msg: "durability disabled"}},
		{`{"type":"replayed","through_seq":42}`, ReplayedFrame{ThroughSeq: 42}},
		{`{"type":"replay_gap","from_seq":1,"earliest_seq":5}`, ReplayGapFrame{FromSeq: 1, EarliestSeq: 5}},
		{`{"type":"lagged","dropped":9}`, LaggedFrame{Dropped: 9}},
		{`{"type":"pong"}`, PongFrame{}},
		{`{"type":"closing","reason":"shutdown"}`, ClosingFrame{Reason: ReasonShutdown}},
		{`{"type":"closing","reason":"revoked"}`, ClosingFrame{Reason: ReasonRevoked}},
	}
	for _, tc := range cases {
		frame, err := parseFrame([]byte(tc.text))
		if err != nil {
			t.Fatalf("parseFrame(%q): %v", tc.text, err)
		}
		if frame != tc.want {
			t.Fatalf("parseFrame(%q) = %#v, want %#v", tc.text, frame, tc.want)
		}
	}
}

// Conformance item 3: subscribe carries replay.after_seq = cursor.
func TestSubscribeMsgUsesCursorAsAfterSeq(t *testing.T) {
	msg := subscribeMsg([]string{"github.>"}, 128, true)
	if msg.Type != "subscribe" || len(msg.Patterns) != 1 || msg.Patterns[0] != "github.>" {
		t.Fatalf("bad subscribe: %+v", msg)
	}
	if msg.Replay == nil || msg.Replay.AfterSeq != 128 {
		t.Fatalf("replay = %+v, want after_seq=128", msg.Replay)
	}
	// Live-only when replay is false.
	live := subscribeMsg([]string{"github.>"}, 128, false)
	if live.Replay != nil {
		t.Fatalf("live-only subscribe should omit replay, got %+v", live.Replay)
	}
}
